### DISCUSSION TITLE
[Question] Rendering into file possible?

### DISCUSSION THREAD

**Discussion by meerfrau** [femtovg/femtovg#167] [Q&A]
Is rendering into a file theoretically possible with femtovg?

**Comment by meerfrau** [comment]
@tronical May you please find some time to reply?

**Comment by tronical** [comment]
Hi! Apologies, I somehow missed this. Yes, in theory that's possible. I guess you would try to create a headless OpenGL context with glutin (with a pbuffer surface), then create your own frame buffer, render into it, and finally read out the pixels. Glutin used to have an example that did this, but it looks like it was removed (the example, not the functionality?).

**Reply by Ionizing** [reply]
Hi, I just wrote an example to do the off-screen rendering with femtovg, but it ended up with totally blank png file. Would it be possible to bother you for some help?

The code is uploaded to https://github.com/Ionizing/femtovg-offscreen

Thanks in advance.

**Reply by tronical** [reply]
Looks good at first glance. By best guess is that after your call to `get_renderer()` the underlying GL surface, context, etc. is dropped (destroyed). Maybe you need to pass those along as well so that they survive the rendering?

**Reply by Ionizing** [reply]
I tried to pass `EventLoop`, `Context` and `OpenGl` object from `get_renderer()` to lengthen the lifetime, but the issue is still there.

**Reply by tronical** [reply]
Oh, I see now: You're creating a surfaceless context (good), but I think you still need to create _some_ surface to render into, i.e. bind the GL framebuffer to an FBO that you've created. Otherwise I think all rendering will go into "nirvana" :)

**Reply by Ionizing** [reply]
Thanks for your advice! I'll try it.
BTW, would it be possible that you make some modification directly to the example and post the code here (or something like pastebins/gist)?

I'm a newbie to graphics, and searched a lot, but the related off-screen OpenGL renderer in rust is rather rare. I looked into the [example from glutin (v0.30)](https://github.com/rust-windowing/glutin/tree/v0.31.0), and it used a home-made renderer and set the framebuffer manually, which means I have no idea on how to tweak femtovg's renderer and set the framebuffers.

**Reply by tronical** [reply]
I'm a little strapped in time, but starting with the glutin example is a good idea. I think what you're missing is [this](https://github.com/rust-windowing/glutin/blob/90d86c5aef0141642862a8240b3408f9410e6059/glutin_examples/examples/egl_device.rs#L77) part in particular, the creation and binding of frame buffers. I suggest to start with that example and verify that it works for you, and then add in the femtovg code.

**Comment by wishawa** [comment]
I got @Ionizing's [femtovg-offscreen example](https://github.com/Ionizing/femtovg-offscreen) to work on on Linux with the `get_renderer()` function below.
The main change is to use a PBuffer surface.

```rust
fn get_renderer() -> OpenGl {
    use std::num::NonZeroU32;

    use glutin::api::egl::device::Device;
    use glutin::api::egl::display::Display;
    use glutin::config::{ConfigSurfaceTypes, ConfigTemplateBuilder};
    use glutin::context::{ContextApi, ContextAttributesBuilder};
    use glutin::prelude::*;
    use glutin::surface::{PbufferSurface, SurfaceAttributesBuilder};

    let devices = Device::query_devices()
        .expect("Failed to query devices")
        .collect::<Vec<_>>();
    for (index, device) in devices.iter().enumerate() {
        println!(
            "Device {}: Name: {} Vendor: {}",
            index,
            device.name().unwrap_or("UNKNOWN"),
            device.vendor().unwrap_or("UNKNOWN")
        );
    }

    let device = devices.first().expect("No available devices");
    let display = unsafe { Display::with_device(device, None) }.expect("Failed to create display");
    let template = ConfigTemplateBuilder::default()
        .with_alpha_size(8)
        .with_surface_type(ConfigSurfaceTypes::PBUFFER)
        .build();

    let config = unsafe { display.find_configs(template) }
        .unwrap()
        .reduce(|config, acc| {
            if config.num_samples() > acc.num_samples() {
                config
            } else {
                acc
            }
        })
        .expect("No available configs");
    println!("Picked a config with {} samples", config.num_samples());

    let context_attributes = ContextAttributesBuilder::new().build(None);
    let fallback_context_attributes = ContextAttributesBuilder::new()
        .with_context_api(ContextApi::Gles(None))
        .build(None);
    let not_current = unsafe {
        display
            .create_context(&config, &context_attributes)
            .unwrap_or_else(|_| {
                display
                    .create_context(&config, &fallback_context_attributes)
                    .expect("failed to create context")
            })
    };
    let surface_attributes = SurfaceAttributesBuilder::<PbufferSurface>::new()
        .with_single_buffer(true)
        .build(NonZeroU32::new(640).unwrap(), NonZeroU32::new(480).unwrap());
    let surf = unsafe { display.create_pbuffer_surface(&config, &surface_attributes) }
        .expect("failed to create surface");
    let _context = not_current.make_current(&surf).unwrap();
    let renderer =
        unsafe { OpenGl::new_from_function_cstr(|s| display.get_proc_address(s) as *const _) }
            .expect("Cannot create renderer");
    return renderer;
}
```

![the image produced](https://github.com/femtovg/femtovg/assets/49518157/82d0766b-b186-4b40-85ed-6f8710634cfc)

I tested this on Fedora with Wayland and an Intel iGPU.
